### ISSUE TITLE
Yolov8 pose struggling to train with a lot of keypoints

### ISSUE THREAD

**Issue by mazatov** [ultralytics/ultralytics#16644]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I'm training a Yolov8 pose model on a custom dataset. I have one object per image and a lot of keypoints in the object. Some visible some not marked with 2 and 0. As I'm training my model, I noticed that the more keypoints I add the more the model struggles to learn. 

So when I get to ~100 keypoints the pose_loss just gets stuck at 12 and doesn't improve. I tried changing some parameters but so far wasn't successful. Any ideas on what to adjust to improve pose learning? The data seems to be annotated correctly as train_batch0 images look OK and if I lower the quantity of keypoints to 30-50 it does start to lower and learn eventually. So the problem seems to stem from the number of keypoints. 

![image](https://github.com/user-attachments/assets/651fd4a7-a88d-49ea-a854-5536c6bc2bd8)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mazatov, thank you for reaching out and for your interest in Ultralytics 🚀! 

For your training challenges with a high number of keypoints, it's great to hear that you've already explored some solutions. Here are some steps you might consider:

### To start, ensure everything is up to date.

Upgrade to the latest `ultralytics` package, with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment using [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

### If this is a 🐛 Bug Report,
Please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to assist us in debugging the issue.

### For further assistance,
- 📘 Visit our [Docs](https://docs.ultralytics.com). The [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) guides are particularly helpful.
- Join our community! Chat with us on [Discord](https://ultralytics.com/discord) 🎧, join discussions on [Discourse](https://community.ultralytics.com), or share insights on our [Subreddit](https://reddit.com/r/ultralytics).

### Running YOLO:

YOLO can be run in a variety of environments including:
- **Notebooks with free GPU:** <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: See the [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon**: See the [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: Visit the [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

### Status Check:

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, our [CI tests](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) pass successfully, verifying YOLO's operation across various platforms.

This is an automated response; one of our Ultralytics engineers will assist you soon. 😊

**Comment by Y-T-G** [comment]
You can try creating adding a P2 layer to create more anchor points. 100 keypoints per object is a lot, so there probably aren't enough anchor points to work with. And you should be using 1 for visible keypoints instead of 2.

**Comment by Y-T-G** [comment]
Related https://github.com/ultralytics/ultralytics/issues/16349

**Comment by mazatov** [comment]
Could you point me in the right direction to look for P2 layer and anchor points? In the issue you provided I see only about the pose loss ` pose: 12.0 # (float) pose loss gain `

**Comment by Y-T-G** [comment]
This is a P2 detection yaml.

https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/models/v8/yolov8-p2.yaml

You will need to create a copy of it and replace the last line:
https://github.com/ultralytics/ultralytics/blob/7d6528a41ac181daee2ef33fcdde5be0d57b0be7/ultralytics/cfg/models/v8/yolov8-p2.yaml#L54
 
and change it to be the same as Pose:

https://github.com/ultralytics/ultralytics/blob/7d6528a41ac181daee2ef33fcdde5be0d57b0be7/ultralytics/cfg/models/v8/yolov8-pose.yaml#L47

except you keep the layer numbers the same as in P2:
`[18, 21, 24, 27]` instead of `[15, 18, 21]`.


Then you just load the model by passing the path to that yaml `YOLO("your_yaml.yaml")` and train normally.

**Comment by Y-T-G** [comment]
> Thanks, what's the difference between 1 and 2 for visibility?

Ultralytics doesn't use 2. It uses 1 for visible keypoints and 0 for invisible keypoints.

**Comment by mazatov** [comment]
Thanks @Y-T-G , Trying this out right now. It's running much slower which I assume is because of the large head size. I also increased the `pose` loss to 215 which is the number of my keypoints so hope that helps too. 

Currently, the model defaults to "n", how can I specify the scale when I run through CLI with this custom yaml? 

![image](https://github.com/user-attachments/assets/30fc3bae-6048-4a7e-981c-a0ea2dc85fe6)

**Comment by Y-T-G** [comment]
You shouldn't increase the pose loss like that. It should be balanced with other losses. So only increase it until it becomes similar to the other losses when the training starts.

The scale is specified based on file name. If you name your yaml file `yolov8m-pose-p2.yaml`, it will use m.

**Comment by mazatov** [comment]
Thanks, didn't know it just took the letter from the file name to pick the scale. Ok, I'll revert to `pose: 12` and try training that way first

**Comment by mazatov** [comment]
@Y-T-G , I seem to be still getting pose loss stuck at 12 even with P2. Do you have any other suggestions? Thanks!

![image](https://github.com/user-attachments/assets/6d741d9c-b91e-4cdc-b2de-d213d74dbd52)

**Comment by mazatov** [comment]
On a separate note, I just read your [blog post on small objects ](https://y-t-g.github.io/tutorials/yolov8-increase-accuracy/) and started wondering as essentially I'm detecting small objects as well. In fact I first tried training an object detector but thought I could obtain better accuracy with pose and that's why started trying it out. My objects have"certain order" like human pose. But I'm not sure how much the pose head of the model cares about the structure of the pose or if it is just detecting them like a standard detector with just distance loss instead of IOU. 

So I'm wondering what's better for labels:
 -  1 Big object and 215 pose keypoints
 -  215 small objects with just one keypoint in their center. 
 - Something in the middle? Have a few large objects with subset of objects on them. 

Do you have any insight into this?

**Comment by Y-T-G** [comment]
Treating them all as objects would be better.

**Comment by mazatov** [comment]
OK, @Y-T-G , I switched my dataset to be of 215 different objects with 1 keypoint in each in the center. I also set up the box and dfl loss to be low. `box=0.01, dfl=0.01`.  However during training all my losses quickly dropped to 0 which makes me think not much learning is going to happen even if I waited for longer. Wonder what can be happening. Thanks for all the insights!


![image](https://github.com/user-attachments/assets/04c4848d-162c-449a-a273-1c634aaf19fa)

**Comment by Y-T-G** [comment]
Try without changing `box`, `dfl` or `pose`

**Comment by mazatov** [comment]
Ok, so I did some more experiments and maybe you can shed some light here. 

1) Without changing `box, dfl, pose` the loss still goes to 0. It looks like the model predicts no objects and that results in loss 0? I'm confused as no objects at all should not yield 0 as I understand. 
2) If I add some other not small objects to the labels than the model starts predicting both large and small objects. It's like it needs some other objects present to not let the model predict nothing. 


On a separate note, does Yolov8/11 differentiate between visibility tag 1 or 2? Some of my keypoints are present in the image, while others are more like ghosts. You can infer their position from overall image and position of other keypoints but there is nothing localized on the image to find them. I was wondering if it would make sense to give "ghost keypoints" visibility 1 and real keypoints visibility 2, but wasn't if Yolov8/11 care for that.

**Comment by Y-T-G** [comment]
Do the batch plots plot them correctly?

**Comment by mazatov** [comment]
Yeah

**Comment by Y-T-G** [comment]
I guess you could try multiplying pose, box and dfl by 100.

**Comment by mazatov** [comment]
@Y-T-G Which loss should be high if all my predictions have very low confidence, like 10e-10?

**Comment by Y-T-G** [comment]
All the losses should be around over 1e-1

**Comment by mazatov** [comment]
@Y-T-G , in that case, I'm still so puzzled 💭 😄 

 - When I have just 215 classes of my small objects represented by small bounding boxes with keypoint in the middle the losses go to 0. The model predicts objects with extremely low confidence, like 10e-10.  
 - When I add one more class=216 of large objects with a keypoint in the middle, it starts to train and detects decently all 216 classes. 

While I found a workaround, I'm still worried that something is not right, if my model can behave like that.

**Comment by Y-T-G** [comment]
How small are your objects?

**Comment by mazatov** [comment]
@Y-T-G , I gave them an arbitrary width and height of 0.02% of the image size. In reality, they are just points, so much smaller.

**Comment by Y-T-G** [comment]
Really small boxes get dropped.

https://github.com/ultralytics/ultralytics/blob/fceaa4003871479e047f607ebb09356a95475e33/ultralytics/data/augment.py#L1271-L1280

**Comment by mazatov** [comment]
They won't be less than 2 pixels. Even when I'm training on 480 pixels, they come out to be 0.02*480 ~ 9 pixels

**Comment by Y-T-G** [comment]
But if you have mosaic augmentation enabled, the images get rescaled, and zoomed in or out.

**Comment by Y-T-G** [comment]
Are you using the P2 Pose model?

**Comment by mazatov** [comment]
No, I've been using default model since I switched to single keypoint objects.

**Comment by Y-T-G** [comment]
Use P2 pose

**Comment by ccl-private** [comment]
![Screenshot from 2024-10-15 10-18-20](https://github.com/user-attachments/assets/3a15bb14-2820-4182-b9dc-43055684d4ed)
@Y-T-G Hi, I was debugging the loss of keypoints. I found that the model directly predicted the absolute position of the point on the picture, and the loss function directly calculated the Euclidean distance between pred and gt. Please tell me, am I understanding this correctly?

**Comment by ccl-private** [comment]
I'm wondering if I can calculate the loss similarly to the DFL loss.

**Comment by Y-T-G** [comment]
It doesn't directly predict the keypoints. That's the decoded version.

https://github.com/ultralytics/ultralytics/blob/15e6133534d82ec6845308f1e258d77c1b9bab36/ultralytics/utils/loss.py#L487

The loss works on the decoded points.

**Comment by ccl-private** [comment]
@Y-T-G Thanks for helping me out.

**Comment by kaspervn** [comment]
I had a problem with similarly small objects. For me this problem became apparent when looking at the `train_batch*.jpg` files it generates at the beginning of the training. Most objects where missing from the previews. I solved it by simply making the boxes bigger ^,^ (but still centered).
